phoenix-core/src/it/java/org/apache/phoenix/end2end/PhoenixRowTimestampFunctionIT.java

```diff
@@ -0,0 +1,398 @@
+                int count = numRows;
+                for (int id = 0; id < count; ++id) {
+                    int idValue = id;
+                    stmt.setInt(1, idValue);
+                    stmt.setDate(2, rowTimestampDate);
+                    stmt.setString(3, "KV1_" + idValue);
+                    stmt.setString(4, "KV2_" + idValue);
+                    stmt.executeUpdate();
+                }
+            }
+            conn.commit();
+        }
+        return tableName;
+    }
+
+    @Test
+    public void testRowTimestampDefault() throws Exception {
+
+        String tableName =  generateUniqueName();
+        try (Connection conn = DriverManager.getConnection(getUrl())) {
+            String ddl = "CREATE TABLE IF NOT EXISTS " + tableName
+                    + " (PK INTEGER NOT NULL PRIMARY KEY, KV1 VARCHAR, KV2 VARCHAR)"
+                    + this.tableDDLOptions;
+            conn.createStatement().execute(ddl);
+
+            String dml = "UPSERT INTO " + tableName + " (PK, KV1, KV2) VALUES (?, ?, ?)";
+            PreparedStatement stmt = conn.prepareStatement(dml);
+
+            int count = 5;
+            for (int id = 0; id < count; ++id) {
+                stmt.setInt(1, id);
+                stmt.setString(2, "KV1_" + id);
+                stmt.setString(3, "KV2_" + id);
+                stmt.executeUpdate();
+            }
+            conn.commit();
+
+            String dql = "SELECT PHOENIX_ROW_TIMESTAMP() FROM " + tableName;
+
+            ResultSet rs = conn.createStatement().executeQuery(dql);
+            // verify row timestamp returned by the query matches the empty column cell timestamp
+            verifyHbaseAllRowsTimestamp(tableName, rs, count);
+
+            // update one row
+            conn.createStatement().execute("UPSERT INTO " + tableName
+                    + " (PK, KV1) VALUES (2, 'KV1_foo')");
+            conn.commit();
+
+            rs = conn.createStatement().executeQuery(dql);
+            // verify again after update
+            verifyHbaseAllRowsTimestamp(tableName, rs, count);
+
+            dql = "SELECT ROWKEY_BYTES_STRING(), PHOENIX_ROW_TIMESTAMP() FROM " + tableName
+                    + " WHERE PK >= 1 AND PK <=3 ";
+            rs = conn.createStatement().executeQuery(dql);
+
+            while (rs.next()) {
+                verifyHbaseRowTimestamp(tableName, rs.getString(1), rs.getDate(2));
+            }
+        }
+    }
+
+    @Test
+    public void testRowTimestampColumn() throws Exception {
+        String tableName =  generateUniqueName();
+        try (Connection conn = DriverManager.getConnection(getUrl())) {
+            String ddl = "CREATE TABLE IF NOT EXISTS " + tableName
+                    + " (PK1 INTEGER NOT NULL, PK2 DATE NOT NULL, KV1 VARCHAR, KV2 VARCHAR"
+                    + " CONSTRAINT PK PRIMARY KEY(PK1, PK2 ROW_TIMESTAMP))" + this.tableDDLOptions;
+            conn.createStatement().execute(ddl);
```

Comment:
Same here. Same comment throughout for all other cases where we don't close the opened statement